### PR TITLE
Cleanup synced products locally on MC disconnect

### LINKED ISSUE #1233: Cleanup synced products locally when disconnecting Merchant Center account
As a followup to #1194 we have the following scenario which is not handled:

1. User connects a merchant account (waits for all the products to sync)
2. User disconnects merchant account (we don't clean up products on disconnect)
3. User connects another merchant account, which triggers a product sync and they are redirected to the product feed page
4. On the product feed page it will fetch all statuses from synced products (which will be all since they were last synced to previous MC account), but they will all get a notFound response (since it's a different MC account)
5. When the product sync runs it will re-sync all the products, but it could take a while before those are available for fetching statuses

The suggestion to resolve this would be to run a cleanup job which would remove the Google ID for all synced products. Since the account is disconnected we would want to mark them unsynced locally without sending a request to remove them in the Merchant Center.

### PR DESCRIPTION
**Pull request by mikkamp**
### Changes proposed in this Pull Request:

This PR resolves the scenario where a Merchant may disconnect their account and then reconnect another account. When they do this all the products are still marked as synced locally until the new product sync has completed. This PR creates a cleanup job that runs when the Merchant Center is disconnected and will mark all products as unsynced. Since it can take a while to complete, I added a check to only continue unsyncing the products as long as the Merchant Center account remains disconnected. Once a new account it reconnected it will stop, any remaining products will later be cleaned up when a new Product Sync happens.

Closes #1233 

### Detailed test instructions:

I added unit testing for the `CleanupSyncedProducts` job, as well as the added `BatchProductHelper` function. However there are no unit tests present for disconnecting the Merchant Center, it was added in the pending PR #1226, once that is merged we can create a followup to confirm the scheduled cleanup is triggered on disconnect.

Steps to test
1. Connect a Merchant account and Sync all products to Google
2. Disconnect the Merchant account, and wait a while till scheduled actions have been run
3. Go to Tools > Scheduled Actions and search for `cleanup_synced`
image
6. Confirm you see all the product IDs that were previously synced (could be spread out across multiple scheduled actions if there is a large amount)

### Changelog entry
* Fix - Cleanup synced products locally when disconnecting Merchant Center account.

### CODE REVIEW

**Comment by mikkamp** [comment]
> However there are no unit tests present for disconnecting the Merchant Center, it was added in the pending PR #1226, once that is merged we can create a followup to confirm the scheduled cleanup is triggered on disconnect.

This has been merged so I created the additional test, see: 8bbf103

File: tests/Unit/Jobs/CleanupSyncedProductsTest.php [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+		$this->action_scheduler->expects( $this->never() )
+			->method( 'schedule_immediate' );
+
+		$this->job->schedule();
+	}
+
+	/**
+	 * Runs before each test is executed.
+	 */
+	public function setUp() {
```
Comment (jorgemd24) [inline tests/Unit/Jobs/CleanupSyncedProductsTest.php]: 💅

I think we can put the setUp function on top of this class so all our tests follow the same structure.

File: src/Jobs/CleanupSyncedProducts.php [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	}
+
+	/**
+	 * Process batch items.
+	 * Skips processing if the Merchant Center has been connected again.
+	 *
+	 * @param int[] $items A single batch of WooCommerce product IDs from the get_batch() method.
+	 */
+	protected function process_items( array $items ) {
+		if ( $this->is_mc_connected() ) {
```
Comment (jorgemd24) [inline src/Jobs/CleanupSyncedProducts.php:63]: 💅

I believe that we do not have any test checking this case. We could process one batch, then connect the mc and try to process the second batch to test that is actually stopping.

**Review by jorgemd24 — APPROVED** [review]
LGTM 👍

I have just left minor comments about the tests.

What I checked:

✅  When I deactivate the MC account the job was scheduled.
✅  The scheduled job had the products ids as an argument. 
✅  The postmeta table was cleaned up.

File: tests/Unit/Jobs/CleanupSyncedProductsTest.php (hunk 1 above)
Comment (mikkamp) [reply tests/Unit/Jobs/CleanupSyncedProductsTest.php]: Moved it. We don't do this in every test though.

File: src/Jobs/CleanupSyncedProducts.php (hunk 2 above)
Comment (mikkamp) [reply src/Jobs/CleanupSyncedProducts.php:63]: The test `test_process_items_with_merchant_center_connected` covers this scenario. It doesn't stop on the second batch, but the code is the same regardless of which batch it processes, so I don't think a test stopping only on the second batch would cover any additional code.

**Comment by mikkamp** [comment]
Thanks for the review and the testing. I've gone ahead and followed up on those comments. I don't think anything extra is needed for that testing scenario, so I'm going to go ahead and merge this.

**Pull request merged**